lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+flexible enough to cover most remapping use-cases, without requiring operators
+to write complex scripts that are difficult to reason about and incur
+a significant performance penalty.
+
+See the [introduction blog post][blog] for more details on the **why**.
+
+[blog]: https://vector.dev/blog/vector-remap-language/#preamble
+
+## Target Audience
+
+VRL has a specialized purpose of solving observability data remapping needs.
+Because of this purpose, the language is mostly used by people who manage
+infrastructure at their organisations. The role of this group is usually
+referred to as "operator", "devops" (Development and Operations) or "SRE" (Site
+Reliability Engineer).
+
+One common generalization of this group is that they are focused on maintaining
+infrastructure within an organization, and often write and maintain their own
+software to achieve this goal. They are adept enough at programming to achieve
+their goals, but have no need or desire to be as skilled in programming as
+dedicated software engineers, because their time is best spent elsewhere.
+
+As with everything, there are exceptions to the rule, and many people in this
+group _are_ highly skilled software engineers, but VRL must capture the largest
+possible segment of this group, and should therefor be limited in complexity.
+
+Therefor, when extending the feature set of VRL, design **the feature for the
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][]
+- classes
+- user-defined functions
+- `goto` statements
+- network calls (see [#4517][])
+
+[#5507]: https://github.com/timberio/vector/issues/5507
+[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
+
+## Conventions
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Use `parse_*` for string decoding functions (e.g. `parse_json` and
+  `parse_grok`).
+
+- Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+
+- Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+- Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+  `format_number`).
```

Comment:
I think we can generalize these and say that functions should be preceded with their function category for organization and discovery. You could explicitly list each, but that feels cumbersome and unnecessary.